### PR TITLE
274 Trim unallocated lines with zero quantity

### LINKED ISSUE #274: Ignore empty placeholder lines when confirming Outbound Shipments
Placeholder lines are created for all items when a user adds a Master List to an Outbound Shipment. 

The user may then only populate details for 1 or 2 items from that list. 

Currently they are then required to manually delete all the remaining placeholder lines before they can proceed. 

Would like to ignore any remaining placeholder lines that have not been edited

(Sorry no screenshots - don't have access to an instance at the mo!)

### PR DESCRIPTION
**Pull request by andreievg**
closes #274 

* Existing check for unallocated lines when status of outbound shipment is changed 'above' new (allocated +), would only check for unallocated lines with non zero quantity
* Zero quantity unallocated lines are trimmed when status changed to to allocated +
* Update useStatusChangePlaceholderCheck to allow placeholder with numberOfPacks === 0

### Tests

- [ ] Create outbound shipment from master list
- [ ] Allocate some lines an leave some unallocated lines as zero
- [ ] Try to change status (lines should be trimmed)

- [ ] Try the same but have unallocated lines with above zero quantity, should see warning

### CODE REVIEW

File: server/service/src/invoice/outbound_shipment/update/generate.rs [hunk 1]
```diff
@@ -70,6 +83,42 @@ pub fn should_update_batches(
     }
 }
 
+// If status changed to allocated and above, remove unallocated lines
+fn unallocated_lines_to_trim(
+    connection: &StorageConnection,
+    invoice: &InvoiceRow,
+    status: &Option<UpdateOutboundShipmentStatus>,
+) -> Result<Option<Vec<InvoiceLineRow>>, RepositoryError> {
+    if invoice.status != InvoiceRowStatus::New {
```
Comment (clemens-msupply) [inline server/service/src/invoice/outbound_shipment/update/generate.rs:92]: nit: comment that there is not status between New and Allocated?

File: server/service/src/invoice/outbound_shipment/update/validate.rs [hunk 2]
```diff
@@ -56,6 +56,8 @@ fn check_invoice_exists(
     Ok(result?)
 }
 
+// If status is beinged changed to allocated and above, return error if there are
```
Comment (clemens-msupply) [inline server/service/src/invoice/outbound_shipment/update/validate.rs:59]:
```suggestion
// If status is changed to allocated and above, return error if there are
```

**Review by clemens-msupply — APPROVED** [review]
Looks good thanks (untested though)

File: server/service/src/invoice/outbound_shipment/update/generate.rs (hunk 1 above)
Comment (andreievg) [reply server/service/src/invoice/outbound_shipment/update/generate.rs:93]: thanks: 708735921a70b25c4750a6d23ca27ab4e85d8ee5

File: server/service/src/invoice/outbound_shipment/update/generate.rs [hunk 3]
```diff
@@ -70,6 +83,43 @@ pub fn should_update_batches(
...
+    let new_invoice_status = match UpdateOutboundShipmentStatus::full_status_option(status) {
+        Some(new_invoice_status) => new_invoice_status,
+        None => return Ok(None),
+    };
+
+    if new_invoice_status == InvoiceRowStatus::New {
+        return Ok(None);
+    }
+
+    // If new invoice status in not new and previous invoice status is new
```
Comment (mark-prins) [inline server/service/src/invoice/outbound_shipment/update/generate.rs:106]: :nit - I think there is an 'in' which should be 'is'? was having trouble following the comment 🤷 
```suggestion
    // If new invoice status is not new and previous invoice status is new
```

**Review by mark-prins — APPROVED** [review]
👍  Nice! works like a charm

File: server/service/src/invoice/outbound_shipment/update/generate.rs (hunk 3 above)
Comment (andreievg) [reply server/service/src/invoice/outbound_shipment/update/generate.rs]: thanks

**Pull request merged**